File: piste.py
"""
Module Piste - Représente une piste d'atterrissage/décollage

LOGIQUE:
Une piste peut être:
- Disponible: Prête à accueillir un avion
- Occupée: Un avion est en train d'atterrir ou de décoller
- En maintenance: Temporairement indisponible

Chaque piste a un numéro unique et peut traiter les atterrissages et décollages
"""

from enum import Enum


class StatutPiste(Enum):
    """Énumération des différents statuts possibles d'une piste"""
    DISPONIBLE = "Disponible"
    OCCUPEE = "Occupée"
    MAINTENANCE = "En maintenance"


class Piste:
    """
    Classe représentant une piste d'atterrissage/décollage
    
    Attributes:
        numero (int): Numéro unique de la piste
        statut (StatutPiste): Statut actuel de la piste
        avion_actuel (Avion): Avion utilisant actuellement la piste (None si disponible)
    """
    
    def __init__(self, numero):
        """
        Initialise une nouvelle piste
        
        Args:
            numero (int): Numéro unique de la piste
        """
        self.numero = numero
        self.statut = StatutPiste.DISPONIBLE
        self.avion_actuel = None
    
    def est_disponible(self):
        """
        Vérifie si la piste est disponible
        
        Returns:
            bool: True si la piste est disponible, False sinon
        """
        return self.statut == StatutPiste.DISPONIBLE
    
    def occuper(self, avion):
        """
        Marque la piste comme occupée par un avion
        
        Args:
            avion (Avion): L'avion qui occupe la piste
            
        Returns:
            bool: True si l'occupation a réussi, False si la piste n'était pas disponible
        """
        if not self.est_disponible():
            return False
        
        self.statut = StatutPiste.OCCUPEE
        self.avion_actuel = avion
        return True
    
    def liberer(self):
        """
        Libère la piste (la rend disponible)
        
        Returns:
            Avion: L'avion qui occupait la piste, ou None
        """
        avion_precedent = self.avion_actuel
        self.statut = StatutPiste.DISPONIBLE
        self.avion_actuel = None
        return avion_precedent
    
    def mettre_en_maintenance(self):
        """Met la piste en maintenance"""
        if self.avion_actuel is None:
            self.statut = StatutPiste.MAINTENANCE
            return True
        return False
    
    def sortir_de_maintenance(self):
        """Sort la piste de maintenance"""
        if self.statut == StatutPiste.MAINTENANCE:
            self.statut = StatutPiste.DISPONIBLE
            return True
        return False
    
    def __str__(self):
        """Représentation textuelle de la piste"""
        avion_info = f"occupée par {self.avion_actuel.code_vol}" if self.avion_actuel else self.statut.value
        return f"Piste {self.numero}: {avion_info}"
    
    def __repr__(self):
        return f"Piste({self.numero})"

File: java/src/Piste.java
/**
 * Piste - Représente une piste d'atterrissage/décollage.
 *
 * LOGIQUE:
 * Une piste est une ressource limitée partagée entre les avions.
 * À tout instant, une piste ne peut être utilisée que par UN SEUL avion.
 *
 * Transitions d'état :
 *   DISPONIBLE → occuper(avion)  → OCCUPEE
 *   OCCUPEE    → liberer()       → DISPONIBLE
 *   DISPONIBLE → mettreEnMaintenance() → MAINTENANCE
 *   MAINTENANCE → sortirDeMaintenance() → DISPONIBLE
 */
public class Piste {

    /** Identifiant unique de la piste */
    private final int numero;

    /** État courant de la piste */
    private StatutPiste statut;

    /** Avion utilisant actuellement la piste (null si disponible) */
    private Avion avionActuel;

    /**
     * Crée une nouvelle piste, initialement disponible.
     *
     * @param numero numéro unique de la piste (commence à 1)
     */
    public Piste(int numero) {
        this.numero = numero;
        this.statut = StatutPiste.DISPONIBLE;
        this.avionActuel = null;
    }

    // ------------------------------------------------------------------ //
    //  Getters
    // ------------------------------------------------------------------ //

    public int getNumero() {
        return numero;
    }

    public StatutPiste getStatut() {
        return statut;
    }

    /** @return l'avion qui occupe la piste, ou null si la piste est libre */
    public Avion getAvionActuel() {
        return avionActuel;
    }

    // ------------------------------------------------------------------ //
    //  Logique métier
    // ------------------------------------------------------------------ //

    /**
     * Vérifie si la piste est disponible.
     *
     * @return true si et seulement si le statut est DISPONIBLE
     */
    public boolean estDisponible() {
        return statut == StatutPiste.DISPONIBLE;
    }

    /**
     * Marque la piste comme occupée par un avion.
     *
     * LOGIQUE:
     *  1. Vérifie que la piste est disponible.
     *  2. Change le statut en OCCUPEE.
     *  3. Mémorise la référence à l'avion.
     *
     * @param avion l'avion qui prend la piste
     * @return true si l'occupation a réussi, false si la piste n'était pas libre
     */
    public boolean occuper(Avion avion) {
        if (!estDisponible()) {
            return false;
        }
        this.statut = StatutPiste.OCCUPEE;
        this.avionActuel = avion;
        return true;
    }

    /**
     * Libère la piste et la rend disponible.
     *
     * @return l'avion qui occupait la piste (peut être null si la piste était déjà libre)
     */
    public Avion liberer() {
        Avion avionPrecedent = this.avionActuel;
        this.statut = StatutPiste.DISPONIBLE;
        this.avionActuel = null;
        return avionPrecedent;
    }

    /**
     * Met la piste en maintenance (seulement si elle est disponible).
     *
     * @return true si la transition a réussi
     */
    public boolean mettreEnMaintenance() {
        if (estDisponible()) {
            this.statut = StatutPiste.MAINTENANCE;
            return true;
        }
        return false;
    }

    /**
     * Sort la piste de maintenance et la rend disponible.
     *
     * @return true si la transition a réussi
     */
    public boolean sortirDeMaintenance() {
        if (statut == StatutPiste.MAINTENANCE) {
            this.statut = StatutPiste.DISPONIBLE;
            return true;
        }
        return false;
    }

    // ------------------------------------------------------------------ //
    //  Affichage
    // ------------------------------------------------------------------ //

    @Override
    public String toString() {
        if (avionActuel != null) {
            return "Piste " + numero + ": occupée par " + avionActuel.getCodeVol();
        }
        return "Piste " + numero + ": " + statut;
    }
}
